backend/pluto/pluto/_internal/adapters/expense_service.py

```diff
@@ -1,12 +1,93 @@
+import csv
+import pathlib
+from typing import List, Tuple
+
 from pluto._internal.domain.model.expense import Expense
 from pluto._internal.domain.ports.database import Database
 from pluto._internal.domain.ports.expense_service import IExpenseService
 
 
+class InvalidRow(Exception):
+    pass
+
+
 class ExpenseServiceImpl(IExpenseService):
     def __init__(self, sm: Database) -> None:
         super().__init__(sm)
 
-    def add_expense(self, expense_dict: dict) -> None:
+    def add_expense_from_dict_without_id(self, expense_dict: dict) -> None:
         expense = Expense.new(**expense_dict)
+        self.add_expense(expense)
+
+    def add_expense(self, expense: Expense):
         self._sm.insert(ExpenseServiceImpl._expense_table, expense.dict())
+
+    def add_expense_from_file(self, file_path: str, user_id: str) -> None:
```

Comment:
Could you please add some unit tests for this method?

Reply:
Sim